leakcanary-android-release/src/main/java/leakcanary/internal/ReleaseHeapAnalyzer.kt

```diff
@@ -0,0 +1,243 @@
+
+import android.app.Application
+import android.content.res.Resources.NotFoundException
+import android.os.Debug
+import android.os.SystemClock
+import leakcanary.HeapAnalysisConfig
+import okio.buffer
+import okio.sink
+import shark.AndroidResourceIdNames
+import shark.CloseableHeapGraph
+import shark.ConstantMemoryMetricsDualSourceProvider
+import shark.HeapAnalysis
+import shark.HeapAnalysisException
+import shark.HeapAnalysisFailure
+import shark.HeapAnalysisSuccess
+import shark.HeapAnalyzer
+import shark.HprofHeapGraph
+import shark.HprofHeapGraph.Companion.openHeapGraph
+import shark.HprofPrimitiveArrayStripper
+import shark.OnAnalysisProgressListener
+import shark.SharkLog
+import java.io.File
+import java.util.UUID
+
+internal class ReleaseHeapAnalyzer(
+  private val application: Application,
+  private val config: HeapAnalysisConfig,
+  private val listener: (HeapAnalysis) -> Unit
+) {
+
+  @Volatile
+  private var stopAnalysing = false
+
+  private val backgroundHandler by lazy {
+    startBackgroundHandlerThread("heap-analyzer")
+  }
+
+  private val analyzeHeap = Runnable {
+    stopAnalysing = false
+    val result = dumpAndAnalyzeHeap()
+    result.heapDumpFile.delete()
+    listener(result)
+  }
+
+  fun removeAllHeapDumpFiles() {
+    backgroundHandler.post {
+      val filesDir = application.filesDir!!
+      val heapDumpFiles = filesDir.listFiles { _, name ->
+        name.startsWith("heap-") && name.endsWith(".hprof")
+      }
+      heapDumpFiles?.forEach { it.delete() }
+    }
+  }
+
+  fun start() {
+    backgroundHandler.removeCallbacks(analyzeHeap)
+    backgroundHandler.post(analyzeHeap)
+  }
+
+  fun stop() {
+    stopAnalysing = true
+    backgroundHandler.removeCallbacks(analyzeHeap)
+  }
+
+  private fun dumpAndAnalyzeHeap(): HeapAnalysis {
+    val filesDir = application.filesDir!!
+    val fileNamePrefix = "heap-${UUID.randomUUID()}"
+    val sensitiveHeapDumpFile = File(filesDir, "$fileNamePrefix.hprof").apply {
+      // Any call to System.exit(0) will run shutdown hooks that will attempt to remove this
+      // file. Note that this is best effort, and won't delete if the VM is killed by the system.
+      deleteOnExit()
+    }
+
+    if (!config.stripHeapDump) {
+      saveResourceIdNamesToMemory()
+    }
+
+    val heapDumpStart = SystemClock.uptimeMillis()
+    saveHeapDumpTime(heapDumpStart)
+
+    var dumpDurationMillis = -1L
+    var analysisDurationMillis = -1L
+    var heapDumpFile = sensitiveHeapDumpFile
+
+    try {
+      dumpHeap(sensitiveHeapDumpFile)
+      dumpDurationMillis = SystemClock.uptimeMillis() - heapDumpStart
+
+      val stripDuration = measureDurationMillis {
+        if (config.stripHeapDump) {
+          val strippedHeapDumpFile = File(filesDir, "$fileNamePrefix-stripped.hprof").apply {
+            deleteOnExit()
+          }
+          heapDumpFile = strippedHeapDumpFile
+          try {
+            stripHeapDump(sensitiveHeapDumpFile, strippedHeapDumpFile)
```

Comment:
Fancy super-nit: `sensitiveHeapDumpFile stripTo strippedHeapDumpFile` method 😄

Reply:
lol not feeling it.